base/compiler/tfuncs.jl

```diff
@@ -853,6 +853,14 @@ function getfield_tfunc(@nospecialize(s00), @nospecialize(name))
     if isa(name, Conditional)
         return Bottom # can't index fields with Bool
     end
+    for _ft in ftypes
+        if (!isa(_ft, Type) && !isa(_ft, TypeVar)) || _ft === Bottom
```

Comment:
What about a type like this:
```
struct Foo
    x
    y::Union{}
    Foo(x) = new(x)
end
```

Reply:
Or similarly this, but this case should possibly be disallowed on construction:
```
struct Bar{T}
    x
    y::T
    Bar{T}(x) where {T} = new(x)
end
Bar{false}(1)
```

Reply:
Is that legal? I guess there's nothing technically wrong with it, but boy. I guess I'll add a check to only bail if it's one of the initialized fields.

Reply:
The first one is definitely legal. The second one probably shouldn't be, since a constructable type needs a layout, and if a field type is a non-type it's impossible to say what its layout is.

Reply:
There's also `s.has_concrete_subtype` which has already checked for conditions such as that?

It's not impossible to describe the layout, but impossible for that field to have a value. We could also treat it similar to apply_type failures, which returns the uninhabited type for the fieldtype and expand the check there (in `inst_ftypes`) instead of here:
```
julia> struct Bar{T}
           x
           y::Complex{T}
           Bar{T}(x) where {T} = new(x)
       end

julia> fieldtypes(Bar{false})
(Any, Union{})
```

Reply:
Yeah, there's no particular necessity in saying the field has type "false", when it's not even valid as a type